File: backend/schemas/content.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class GenerationRequest(BaseModel):
    topic: str = Field(min_length=1, max_length=500)
    platform: str = Field(default="linkedin", max_length=32)
    tone: str = Field(default="professional", max_length=32)
    max_length: int = Field(default=300, ge=50, le=3000)


class PipelineGenerateRequest(BaseModel):
    topic: str | None = Field(default=None, max_length=500)
    platform: str = Field(default="linkedin", max_length=32)
    tone: str = Field(default="professional", max_length=32)
    max_length: int = Field(default=300, ge=50, le=3000)
    pipeline_type: Literal["daily_brief", "draft_generation", "analytics"] = "draft_generation"


class PipelineStatusResponse(BaseModel):
    pipeline_id: str
    is_complete: bool
    current_step: str
    steps_completed: list[str]
    errors: list[dict[str, Any]]
    duration_ms: int
    requires_human_approval: bool = False


class GeneratedPostResponse(BaseModel):
    id: str
    title: str
    body: str
    hook: str | None = None
    call_to_action: str | None = None
    hashtags: list[str] = []
    quality_score: float | None = None
    review_feedback: str | None = None
    status: str = "draft"
    platform: str = "linkedin"
    created_at: str
    published_at: str | None = None


class ReviewResponse(BaseModel):
    verdict: str
    score: float
    feedback: str
    issues: list[str] = []


class IdeasRequest(BaseModel):
    topic: str | None = Field(default=None, max_length=500)
    platform: str = Field(default="linkedin", max_length=32)
    count: int = Field(default=5, ge=1, le=20)


class IdeaResponse(BaseModel):
    id: str
    title: str
    description: str
    relevance_score: float = 0.0
    platform: str = "linkedin"


class EvaluateContentRequest(BaseModel):
    title: str = Field(min_length=1, max_length=500)
    body: str = Field(min_length=1, max_length=10000)
    platform: str = Field(default="linkedin", max_length=32)


class EvaluateContentResponse(BaseModel):
    score: float
    verdict: str
    feedback: str
    issues: list[str]
    strengths: list[str]


class PipelineOutputResponse(BaseModel):
    pipeline_id: str
    pipeline_type: str
    total_duration_ms: int
    steps_completed: int
    errors: int
    requires_human_approval: bool
    draft: dict[str, Any] = {}
    review: dict[str, Any] = {}
    strategy: dict[str, Any] = {}
    topic: str = ""


class HistoryListResponse(BaseModel):
    records: list[GeneratedPostResponse]
    total: int
    page: int
    page_size: int

File: frontend/lib/api/content.ts
import { apiGet, apiPatch, apiPost } from "./client";

export interface PipelineRequest {
  topic?: string;
  platform?: string;
  tone?: string;
  max_length?: number;
  pipeline_type?: "daily_brief" | "draft_generation" | "analytics";
}

export interface PipelineStatus {
  pipeline_id: string;
  is_complete: boolean;
  current_step: string;
  steps_completed: string[];
  errors: { step: string; error: string }[];
  duration_ms: number;
  requires_human_approval: boolean;
}

export interface IdeasRequest {
  topic?: string;
  platform?: string;
  count?: number;
}

export interface Idea {
  id: string;
  title: string;
  description: string;
  relevance_score: number;
  platform: string;
}

export interface EvaluateRequest {
  title: string;
  body: string;
  platform?: string;
}

export interface EvaluateResponse {
  score: number;
  verdict: string;
  feedback: string;
  issues: string[];
  strengths: string[];
}

export interface GeneratedPost {
  id: string;
  title: string;
  body: string;
  hook: string | null;
  call_to_action: string | null;
  hashtags: string[];
  quality_score: number | null;
  review_feedback: string | null;
  status: "draft" | "published" | "archived";
  platform: string;
  created_at: string;
  published_at: string | null;
}

export interface HistoryResponse {
  records: GeneratedPost[];
  total: number;
  page: number;
  page_size: number;
}

export async function generateContent(
  request: PipelineRequest,
): Promise<PipelineStatus> {
  return apiPost<PipelineStatus>("/content/generate", request);
}

export async function getPipelineStatus(
  pipelineId: string,
): Promise<PipelineStatus> {
  return apiGet<PipelineStatus>(`/content/pipeline/${pipelineId}`);
}

export async function getPipelineOutput(
  pipelineId: string,
): Promise<Record<string, unknown>> {
  return apiGet(`/content/pipeline/${pipelineId}/output`);
}

export async function generateIdeas(
  request: IdeasRequest,
): Promise<Idea[]> {
  return apiPost<Idea[]>("/content/ideas", request);
}

export async function evaluateContent(
  request: EvaluateRequest,
): Promise<EvaluateResponse> {
  return apiPost<EvaluateResponse>("/content/evaluate", request);
}

export async function getHistory(params: {
  page?: number;
  page_size?: number;
  platform?: string;
  status?: string;
}): Promise<HistoryResponse> {
  return apiGet<HistoryResponse>("/content/history", params as Record<string, string | number | undefined>);
}

export async function getHistoryRecord(
  recordId: string,
): Promise<GeneratedPost> {
  return apiGet<GeneratedPost>(`/content/history/${recordId}`);
}

export async function updateRecordStatus(
  recordId: string,
  status: "draft" | "published" | "archived",
): Promise<GeneratedPost> {
  return apiPatch<GeneratedPost>(
    `/content/history/${recordId}/status?status=${status}`,
  );
}
